### ISSUE TITLE
grdimage does not plot subsets of data correctly on a global map

### ISSUE THREAD

**Issue by MarkWieczorek** [GenericMappingTools/pygmt#732]
# Problem

I have a subset of data that I would like to plot on a global projection (such as mollweide). This works with GMT, but pygmt provides incorrect output.

# Example

Let's create a global dataset of random numbers on a grid with 1 degree spacing in latitude and longitude:

```
lats = np.arange(90., -91., -1.)
lons = np.arange(0., 361., 1.)
data = np.random.rand(181, 361)
da = xr.DataArray(data, coords=[('latitude', lats), ('longitude', lons)])
```

Now let's plot only the region from 90N-30S and 0E-220E with a Mollweide projection centered on 180 E
```
figure = pygmt.Figure()
pygmt.makecpt(series=[0., 1., 0.1], cmap='vik', continuous=True)
figure.grdimage(da[0:121, 0:221], region='g', projection='W180/8i', frame=True)
figure.show(method='external')
```
![ex1](https://user-images.githubusercontent.com/8697285/102112424-f46f7280-3e37-11eb-8290-a8902bbad750.png)

This is as expected, but now do the same, but centered on 0 E
```
figure = pygmt.Figure()
pygmt.makecpt(series=[0., 1., 0.1], cmap='vik', continuous=True)
figure.grdimage(da[0:121, 0:221], region='g', projection='W0/8i', frame=True)
figure.show(method='external')
```

![ex2](https://user-images.githubusercontent.com/8697285/102112833-8f684c80-3e38-11eb-91bd-aed9219c30b2.png)

And as a second example, here I want to plot the global data, but using a dataset the doesn't contain the redundant data at 360 E

```
figure = pygmt.Figure()
pygmt.makecpt(series=[0., 1., 0.1], cmap='vik', continuous=True)
figure.grdimage(da[0:181, 0:360], region='g', projection='W0/8i', frame=True)
figure.show(method='external')
```
![ex3](https://user-images.githubusercontent.com/8697285/102112498-0ea95080-3e38-11eb-8982-22e567908899.png)
This is a complete failure, but changing the central meridian to 180 degrees works!
```
figure = pygmt.Figure()
pygmt.makecpt(series=[0., 1., 0.1], cmap='vik', continuous=True)
figure.grdimage(da[0:181, 0:360], region='g', projection='W180/8i', frame=True)
figure.show(method='external')
```
![ex4](https://user-images.githubusercontent.com/8697285/102112511-149f3180-3e38-11eb-9848-f6ea9a7b29ee.png)

**Comment by seisman** [comment]
@PaulWessel I think this is also an upstream GMT bug, possibly related to https://github.com/GenericMappingTools/gmt/issues/4335. Please see if you can help.

To reproduce the issue using the above codes, you need to import the following packages first:
```
import pygmt
import numpy as np
import xarray as xr
```

**Comment by seisman** [comment]
Ping @PaulWessel

**Comment by weiji14** [comment]
Still an issue on GMT 6.2.0rc1

```python
import numpy as np
import pygmt
import xarray as xr

lats = np.arange(90.0, -91.0, -1.0)
lons = np.arange(0.0, 361.0, 1.0)
data = np.random.rand(181, 361)
da = xr.DataArray(data, coords=[("latitude", lats), ("longitude", lons)])

fig = pygmt.Figure()
pygmt.makecpt(series=[0.0, 1.0, 0.1], cmap="vik", continuous=True)
fig.grdimage(da[0:121, 0:221], region="g", projection="W0/8i", frame=True)
fig.savefig("temp.png")
fig.show()
```

![temp](https://user-images.githubusercontent.com/23487320/115815135-84f33a80-a44a-11eb-9ea4-0f58488bdb65.png)

**Comment by PaulWessel** [comment]
Busy with taxes this weekend but will get on this - sorry it slipped off the radar.  Seems like I fixed a very related problem not that long ago, no @seisman ?

**Comment by PaulWessel** [comment]
Not sure if https://docs.generic-mapping-tools.org/dev/debug.html#debug-pygmt-in-xcode-on-macos has all required steps.  After git pull, make install, conda activate pygmt I try (from pygmt directory)

```
python ~/bad.py (this is the script posted above)
Traceback (most recent call last):
  File "/Users/pwessel/bad.py", line 2, in <module>
    import pygmt
  File "/Users/pwessel/GMTdev/pygmt/pygmt/__init__.py", line 28, in <module>
    from pygmt.figure import Figure, set_display
  File "/Users/pwessel/GMTdev/pygmt/pygmt/figure.py", line 9, in <module>
    import IPython
ModuleNotFoundError: No module named 'IPython'
```

**Comment by seisman** [comment]
> Not sure if [docs.generic-mapping-tools.org/dev/debug.html#debug-pygmt-in-xcode-on-macos](https://docs.generic-mapping-tools.org/dev/debug.html#debug-pygmt-in-xcode-on-macos) has all required steps. After git pull, make install, conda activate pygmt I try (from pygmt directory)
> 
> ```
> python ~/bad.py (this is the script posted above)
> Traceback (most recent call last):
>   File "/Users/pwessel/bad.py", line 2, in <module>
>     import pygmt
>   File "/Users/pwessel/GMTdev/pygmt/pygmt/__init__.py", line 28, in <module>
>     from pygmt.figure import Figure, set_display
>   File "/Users/pwessel/GMTdev/pygmt/pygmt/figure.py", line 9, in <module>
>     import IPython
> ModuleNotFoundError: No module named 'IPython'
> ```

I think you found a bug which was introduced recently. 

Your temporary fix is:
```
conda activate python
conda install ipython
```

**Comment by PaulWessel** [comment]
So I am on miniconda:

(pygmt) pwessel@macnut:~/GMTdev/pygmt-> conda activate python
Could not find conda environment: python
You can list all discoverable environments with `conda info --envs`.

**Comment by seisman** [comment]
Sorry, typo. I meant:

```
conda activate pygmt
conda install ipython
```

**Comment by PaulWessel** [comment]
OK, that installed a bunch of stuff including ipython.  But still some PATH issues perhaps (remember, I do this so infrequently I may not have things set up right):

Traceback (most recent call last):
  File "/Users/pwessel/bad.py", line 2, in <module>
    import pygmt
  File "/Users/pwessel/GMTdev/pygmt/pygmt/__init__.py", line 28, in <module>
    from pygmt.figure import Figure, set_display
  File "/Users/pwessel/GMTdev/pygmt/pygmt/figure.py", line 9, in <module>
    import IPython
ModuleNotFoundError: No module named 'IPython'

**Comment by seisman** [comment]
What's the output of `which python`? Just want to make sure that you're running the one in the pygmt environment.

**Comment by PaulWessel** [comment]
/Users/pwessel/opt/miniconda3/bin/python

**Comment by seisman** [comment]
The path is incorrect. After activating the pygmt environment, you should be using `/Users/pwessel/opt/miniconda3/envs/pygmt/bin/python`. Did you forget to run `conda activate pygmt`?

**Comment by PaulWessel** [comment]
No, and just run it again.  Same path coming back.

**Comment by seisman** [comment]
It doesn't make sense to me.

At line 10 of file `/Users/pwessel/GMTdev/pygmt/pygmt/figure.py`, changing `KeyError` to `ModuleNotFoundError`, then try to run the bad.py script, does it work for you?

**Comment by PaulWessel** [comment]
That runs fine.  I get a ton of debug memory leak messages that should be fixed in general but the script works and produces the failed image.

**Comment by weiji14** [comment]
Ok, some testing on GMT 6.2.0rc2 (install using `conda install -c conda-forge/label/dev gmt=6.2.0rc2`):

```python
# Random data setup
import numpy as np
import pygmt
import xarray as xr

lats = np.arange(90.0, -91.0, -1.0)
lons = np.arange(0.0, 361.0, 1.0)
data = np.random.rand(181, 361)
da = xr.DataArray(data, coords=[("latitude", lats), ("longitude", lons)])
```

### Case 1

Previous whitespaces on the left and right are now coloured in correctly.

```python
fig = pygmt.Figure()
pygmt.makecpt(series=[0.0, 1.0, 0.1], cmap="vik", continuous=True)
fig.grdimage(da[0:121, 0:221], region="g", projection="W0/8i", frame=True)
fig.savefig("temp1.png")
fig.show()
```

![temp1](https://user-images.githubusercontent.com/23487320/119580102-e45eb480-be13-11eb-85ee-432620ca1d72.png)

### Case 2

This previously failed with a complete blue image. Now the random data shows up, but there is still a white line running through the middle 

```python
figure = pygmt.Figure()
pygmt.makecpt(series=[0., 1., 0.1], cmap='vik', continuous=True)
figure.grdimage(da[0:181, 0:360], region='g', projection='W0/8i', frame=True)
figure.savefig("temp2.png")
figure.show()
```

![temp2](https://user-images.githubusercontent.com/23487320/119580140-f5a7c100-be13-11eb-9292-6950dddca9d3.png)

**Comment by PaulWessel** [comment]
You go 0 to 181 in latitude but not 0 to 361 in longitude?  Aren't you missing a column?

**Comment by weiji14** [comment]
> You go 0 to 181 in latitude but not 0 to 361 in longitude? Aren't you missing a column?

Ah right, thanks Paul. So we do need a redundant column to wrap around the dateline. Using `figure.grdimage(da[0:181, 0:361], region="g", projection="W0/8i", frame=True)` removes the middle white line:

![temp3](https://user-images.githubusercontent.com/23487320/119589211-740d5e80-be26-11eb-961e-a6f4d84f4c40.png)

Ok to close this issue @seisman since it's resolved with GMT 6.2.0rc2? Or do we want to do some more work/testing on this?

**Comment by seisman** [comment]
> Ok to close this issue @seisman since it's resolved with GMT 6.2.0rc2? Or do we want to do some more work/testing on this?

I think we need to add a test for it although it's an upstream bug.

**Comment by weiji14** [comment]
Just a note that doing a slice of an `xarray.DataArray` grid resets the coordinate system from Geographic type to Cartesian type as mentioned in https://github.com/GenericMappingTools/pygmt/issues/524#issuecomment-662274140.
